File: Linked Lists/Palindrome Linked list/palindrome_ll.java
class ListNode {
    int val;
    ListNode next;

    ListNode(int val) {
        this.val = val;
        this.next = null;
    }
}

public class palindrome_ll {

    public boolean isPalindrome(ListNode head) {
        if (head == null || head.next == null) {
            return true;
        }

       
        ListNode slow = head;
        ListNode fast = head;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        ListNode prev = null;
        ListNode curr = slow;
        while (curr != null) {
            ListNode nextNode = curr.next;
            curr.next = prev;
            prev = curr;
            curr = nextNode;
        }

        
        ListNode first = head;
        ListNode second = prev;
        while (second != null) {
            if (first.val != second.val) {
                return false;
            }
            first = first.next;
            second = second.next;
        }

        return true;
    }

    
    public static void printList(ListNode head) {
        ListNode temp = head;
        while (temp != null) {
            System.out.print(temp.val);
            if (temp.next != null) System.out.print(" -> ");
            temp = temp.next;
        }
        System.out.println();
    }

  
    public static void main(String[] args) {
     
        ListNode head1 = new ListNode(1);
        head1.next = new ListNode(2);
        head1.next.next = new ListNode(2);
        head1.next.next.next = new ListNode(1);

        System.out.print("Testing list: ");
        printList(head1);

        palindrome_ll sol = new palindrome_ll();
        System.out.println("Is Palindrome? " + sol.isPalindrome(head1));

        ListNode head2 = new ListNode(1);
        head2.next = new ListNode(2);
        head2.next.next = new ListNode(3);

        System.out.print("\nTesting list: ");
        printList(head2);
        System.out.println("Is Palindrome? " + sol.isPalindrome(head2));
    }
}

File: Linked Lists/Palindrome Linked list/palindrome_ll.py
from typing import Optional

# Node definition
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        if head is None or head.next is None:
            print("Single node or empty list — palindrome by default.")
            return True

        slow = head
        fast = head
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
            
        prev = None
        curr = slow
        while curr:
            nxt = curr.next
            curr.next = prev
            prev = curr
            curr = nxt

        # Step 3: Compare first and second halves
        first = head
        second = prev
        while second and first:
            # Debug print
            print(f"Comparing {first.val} and {second.val}")

            if first.val != second.val:
                print("Mismatch found — Not a palindrome.")
                return False

            first = first.next
            second = second.next

        print("All matched — Palindrome confirmed.")
        return True


# --- MAIN FUNCTION ---
def main():
    # ✅ Example 1: Palindrome list
    head = ListNode(1)
    head.next = ListNode(2)
    head.next.next = ListNode(2)
    head.next.next.next = ListNode(1)

    sol = Solution()
    print("Testing linked list: 1 -> 2 -> 2 -> 1")
    print("Result:", sol.isPalindrome(head))

    # ✅ Example 2: Not a palindrome
    head2 = ListNode(1)
    head2.next = ListNode(2)
    head2.next.next = ListNode(3)

    print("\nTesting linked list: 1 -> 2 -> 3")
    print("Result:", sol.isPalindrome(head2))


if __name__ == "__main__":
    main()
